### PR TITLE
WIP: Fix several issues in realtime-compositions

### PR DESCRIPTION
**Pull request by sttts**
### Description of your changes

In collaboration with @haarchri, fixing a number of issues in realtime-compositions. In particular, ensure the right life-cycle of watches and informers to match that of the composite controller:

- ~~apiextensions/definition: don't attempt to start multiple composite controllers~~ merged through https://github.com/crossplane/crossplane/pull/5437
- apiextensions/definition: fix life-cycle of composite informers to be that of the composite controller
- ~~apiextensions/definition: fix leak of watch on composition revisions due to wrong referenced cache~~ moved to https://github.com/crossplane/crossplane/pull/5463
- ~~apiextensions/definition: unregister composite gvk from composedResourceInformers everywhere~~ merged through https://github.com/crossplane/crossplane/pull/5437
- apiextensions/definition: wait for sync in composedResourceInformers and delay cd informer deletion
- revert https://github.com/crossplane/crossplane/pull/5296.

~~Depends on https://github.com/crossplane/crossplane-runtime/pull/672.~~

Fixes https://github.com/crossplane/crossplane/issues/5400

I have: 

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- ~~[ ] Added or updated unit tests.~~
- [x] Added or updated e2e tests.
- ~~[ ] Linked a PR or a [docs tracking issue] to [document this change].~~
- ~~[ ] Added `backport release-x.y` labels to auto-backport this PR.~~

Need help with this checklist? See the [cheat sheet].

[contribution process]: https://github.com/crossplane/crossplane/tree/master/contributing
[docs tracking issue]: https://github.com/crossplane/docs/issues/new
[document this change]: https://docs.crossplane.io/contribute/contribute
[cheat sheet]: https://github.com/crossplane/crossplane/tree/master/contributing#checklist-cheat-sheet

### CODE REVIEW

File: internal/controller/apiextensions/definition/reconciler.go [hunk 1]
```diff
@@ -471,19 +470,27 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
...
+		//
+		// Note: we explicitly use the xrCache for composition revisions to sync
+		// the life-cycle of the source and the composite list, and to avoid
+		// leaking these triggers (source.Kind does not remove a handler ever).
+		controller.TriggeredBy(&lazySyncingSource{Create: func() source.SyncingSource {
+			return source.Kind(xrCache, &v1.CompositionRevision{})
+		}}, handler.Funcs{
+			CreateFunc: composite.EnqueueForCompositionRevisionFunc(ck, func(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error {
+				return xrCache.List(ctx, list, opts...)
+			}, r.log),
```
Comment (negz) [inline internal/controller/apiextensions/definition/reconciler.go:486]: Style nit: This is a bit of a handful to read due to the nested function definitions with long signatures.

File: internal/controller/apiextensions/composite/reconciler.go [hunk 2]
```diff
@@ -712,6 +712,7 @@ func EnqueueForCompositionRevisionFunc(of resource.CompositeKind, list func(ctx
 			log.Info("cannot list in CompositionRevision handler", "type", schema.GroupVersionKind(of).String(), "error", err)
 			return
 		}
+		log.Debug("Enqueueing composite resources because a new CompositionRevision was created", "count", len(xrs.Items))
```
[the remark this answers is not in the document]
Comment (sttts) [reply internal/controller/apiextensions/composite/reconciler.go:715]: done

File: internal/controller/apiextensions/definition/composed.go [hunk 3]
```diff
@@ -105,15 +116,34 @@ func (i *composedResourceInformers) Start(ctx context.Context, h handler.EventHa
...
+
+	// asynchronously mark the cache as synced. We won't list XRs before this
+	// happens. We delay composite informer deletion by at least
+	// minTimeToKeepAlive such that there is time to sync.
+	go func() {
+		if ca.WaitForCacheSync(context.Background()) {
+			i.lock.Lock()
+			defer i.lock.Unlock()
+			if _, ok := i.xrCaches[gvk]; ok {
+				i.log.Debug("Composite resource cache seen synced", "gvk", gvk.String(), "after", time.Since(startAt))
```
[the remark this answers is not in the document]
Comment (sttts) [reply internal/controller/apiextensions/definition/composed.go:139]: have switched to gvk.

File: internal/controller/apiextensions/definition/composed.go [hunk 4]
```diff
@@ -105,15 +116,34 @@ func (i *composedResourceInformers) Start(ctx context.Context, h handler.EventHa
...
 	defer i.lock.Unlock()
 
+	startAt := time.Now()
+
 	if i.xrCaches == nil {
 		i.xrCaches = make(map[schema.GroupVersionKind]cache.Cache)
 	}
 	i.xrCaches[gvk] = ca
+
+	// asynchronously mark the cache as synced. We won't list XRs before this
```
[the remark this answers is not in the document]
Comment (sttts) [reply internal/controller/apiextensions/definition/composed.go:131]: done

File: internal/controller/apiextensions/definition/reconciler.go (hunk 1 above)
Comment (sttts) [reply internal/controller/apiextensions/definition/reconciler.go:486]: have made it similar to the other source. A little bit easier to read. The closured for delaying evaluation are little ugly. But Go doesn't allow a better syntax. Wished there was a `foo(bla, lazy xrCache, ...)`.

**Comment by jbw976** [comment]
Looks like confidence in this approach is growing as per https://github.com/crossplane/crossplane/issues/5400#issuecomment-1986420180. I've added the backport labels for v1.14 and v1.15 so once this is merged we can likely get patch releases out with this fix 💪

**Comment by jbw976** [comment]
ah and this one may be the most impactful of the bunch @sttts if you're able to make progress on this for v1.16. Sorry for the multiple notifications, just going through all the open PRs now that could be impactful for v1.16 😉

**Comment by haarchri** [comment]
this part https://github.com/crossplane/crossplane/issues/5400 is still open - which gets fixed here - so i wonder if we can cherry pick this

**Comment by jbw976** [comment]
Thanks for the reminder on this PR and that you're still seeing issues with realtime compositions even after the big changes in https://github.com/crossplane/crossplane/pull/5651 @haarchri!

I'll make sure to peruse through the changes in this PR and work with @sttts and @negz to see if they can be ported to the state of the world after https://github.com/crossplane/crossplane/pull/5651.  This PR can't be re-opened since GitHub says the `master` branch has been deleted, but I've added it to the [beta tracking epic](https://github.com/crossplane/crossplane/issues/4828) as a reminder to look more deeply soon 🙇‍♂️

**Pull request closed without merging**